### PR TITLE
Introduce VirtualIdEmbeddable and IdClassEmbeddable

### PR DESCRIPTION
**Pull request by sebersole**
Introduce VirtualIdEmbeddable and IdClassEmbeddable

### CODE REVIEW

**Comment by sebersole** [comment]
Ongoing work, but in a working state and I wanted to get it pushed before the weekend :)

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/IdClassEmbeddable.java [hunk 1]
```diff
@@ -0,0 +1,780 @@
...
+				creationProcess
+		);
+	}
+
+
+
+
+
+	// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+	// init
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/IdClassEmbeddable.java:490]: I think we should try to move most of this logic into some abstract base class if possible, as that is AFAICS duplicated from EmbeddableMappingTypeImpl. It appears that the same/similar duplicate code is also used in VirtualIdEmbeddable

**Review by beikov — COMMENTED** [review]
This looks mostly like a great refactoring, although there are some rough edges I mentioned, I think this is a nice step forward!

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/IdClassEmbeddable.java (hunk 1 above)
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/IdClassEmbeddable.java:490]: Yep, as mentioned in my commit comments ;)

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/MappingModelCreationHelper.java [hunk 2]
```diff
@@ -161,78 +153,88 @@ public static CompositeIdentifierMapping buildNonEncapsulatedCompositeIdentifier
...
+//
+//		// Inject the model part also in the composite type of the id-class, because that is what we actually "instantiate"
+//		// which needs the model part for instantiation
+////		final CompositeIdentifierMapping compositeIdentifierMapping = (CompositeIdentifierMapping) virtualIdEmbeddable.getEmbeddedValueMapping();
+////		( (CompositeTypeImplementor) virtualIdSource.getType() ).injectMappingModelPart(
+////				(EmbeddableValuedModelPart) compositeIdentifierMapping,
+////				creationProcess
+////		);
+//
+//		return compositeIdentifierMapping;
```
[the remark this answers is not in the document]
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/MappingModelCreationHelper.java:237]: Yep, as mentioned in my commit comments this is still a work in progress

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/NonAggregatedIdentifierMappingImpl.java [hunk 3]
```diff
@@ -46,44 +47,105 @@
  * @author Steve Ebersole
  * @apiNote Technically a MapsId id does not have to be composite; we still handle that this class however
  */
-public class NonAggregatedIdentifierMappingImpl extends AbstractCompositeIdentifierMapping {
+public class NonAggregatedIdentifierMappingImpl extends AbstractCompositeIdentifierMapping implements NonAggregatedIdentifierMapping {
+	private final VirtualIdEmbeddable virtualIdEmbeddable;
+	private final IdClassEmbeddable idClassEmbeddable;
```
[the remark this answers is not in the document]
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/NonAggregatedIdentifierMappingImpl.java:52]: The id-mapping is the embedded, not the embeddable.

File: hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableInstantiatorPojoStandard.java [hunk 4]
```diff
@@ -54,29 +50,29 @@ public EmbeddableInstantiatorPojoStandard(
...
+			throw new InstantiationException( "Unable to locate constructor for embeddable", getMappedPojoClass() );
 		}
 
-		if ( valuesAccess != null ) {
-			if ( constructor.getParameterTypes().length > 0 ) {
-				// constructor injection
-				throw new NotYetImplementedFor6Exception( "Constructor injection for embeddables not yet implemented" );
+		try {
+			if ( constructorInjection ) {
+				return constructor.newInstance( valuesAccess.get() );
```
[the remark this answers is not in the document]
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/metamodel/internal/EmbeddableInstantiatorPojoStandard.java:64]: Well considering constructor injection has not actually been integrated yet we really aren't sure what this block will look like...

File: hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/NonAggregatedIdentifierMappingImpl.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (sebersole) [reply hibernate-core/src/main/java/org/hibernate/metamodel/mapping/internal/NonAggregatedIdentifierMappingImpl.java:52]: You mean the id-class I guess... I prefer that to be explicit.

**Pull request closed without merging**